### PR TITLE
Fix modin pytests

### PR DESCRIPTION
**Pull request by igalink**
In order to fix modin pytests:
 unic_ptr was replaced by shared_ptr and origin recordset stored on C++ side.

### CODE REVIEW

File: Embedded/DBEngine.cpp [hunk 1]
```diff
@@ -494,6 +497,7 @@ class DBEngineImpl : public DBEngine {
   Catalog_Namespace::UserMetadata user_;
   bool is_temp_db_;
   std::string udf_filename_;
+std::shared_ptr<CursorImpl> cursor_;
```
Comment (anton-malakhov) [inline Embedded/DBEngine.cpp:500]: broken formatting

**Comment by fexolm** [comment]
It's not very important, but I'm just curious, what issue does shared_ptr solves?

**Review by anton-malakhov — CHANGES_REQUESTED** [review]
LGTM, however, there are few questions.
This PR lacks informative description of what pytest errors were fixed and why did you change unique to shared?
Does it need additional tests? Which tests were run with these changes? I'd recommend running our modin tests in order to be sure that this new semantics does not break our prototypes

**Comment by igalink** [comment]
test_dataframe.py: 12 failed, 168 passed, 262 warnings in 25.22s 
TestCSV::test_csv_fillna - AssertionError: Series are different
TestMasks::test_iloc - AssertionError: numpy array are different
TestConcat::test_insert - AssertionError: numpy array are different
TestGroupby::test_df_astype - AssertionError: numpy array are different
TestGroupby::test_h2o_q7 - AssertionError: numpy array are different
TestMerge::test_h2o_q4 - AssertionError: numpy array are different
TestMerge::test_h2o_q5 - AssertionError: numpy array are different
TestBinaryOp::test_add_list - AssertionError: numpy array are different
TestBinaryOp::test_add_columns_and_assign_to_existing - AssertionError: numpy arr...
TestBinaryOp::test_mul_cst - AssertionError: numpy array are different
TestBinaryOp::test_floordiv_cst - AssertionError: numpy array are different
TestBinaryOp::test_floordiv_list - AssertionError: numpy array are different

**Comment by igalink** [comment]
All above errors appeared because of commit:
SHA-1: 83e997931a2da8c868bcb069d975718b279b8e28
( Fix memory leaks in omnisci embedded engine (#66))

Rootcause is corrupted recordset in the python. Fixed by managing cursor destruction on the c++ side (cursor_).

**Comment by igalink** [comment]
> It's not very important, but I'm just curious, what issue does shared_ptr solve?

shared_ptr - just a tool to share cursor between c++ and python. Fix is the DBEngineImpl::cursor_ , which guarantees that the object is deleted exclusively on the C++ side without using a Python garbage collector.

**Comment by igalink** [comment]
Test results after fix: pytest modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py
180 passed, 262 warnings in 20.03s

**Comment by fexolm** [comment]
It looks like you don't use shared_ptr properly. As I see, you just want to make DBEngine own the cursor. This way you should use unique_ptr to store cursor_, and return references to Cursor objects in API methods. However, as far as I see, your solution would restrict having multiple cursors simultaneously. 
What is the problem in allowing python to own the cursors?

**Comment by fexolm** [comment]
I mean, in the current solution, python would still own the Cursor and if you call reset, the cursor would be destructed on the python side. Isn't this a behaviour you would like to avoid for whatever reason?

**Comment by igalink** [comment]
> It looks like you don't use shared_ptr properly. As I see, you just want to make DBEngine own the cursor. This way you should use unique_ptr to store cursor_, and return references to Cursor objects in API methods. However, as far as I see, your solution would forbid having multiple cursors simultaneously.

I decided to use managed object on python side also (shared_ptr instead of c++ pointer). Why is it not proper?
Practice has shown that there is no need for multiple cursors.

**Comment by igalink** [comment]
> I mean, in the current solution, python would still own the Cursor and if you call reset, the cursor would be destructed on the python side. Isn't this a behaviour you would like to avoid for whatever reason?

Theoretically, this is so, but in practice we have a destroyed/corrupted cursor in tests.

File: Embedded/DBEngine.cpp (hunk 1 above)
Comment (igalink) [reply Embedded/DBEngine.cpp]: Fixed

**Comment by ienkovich** [comment]
Previously we didn't destroy cursors at all and the recent patch introduced their destruction. This revealed some problems with data lifetime we don't understand. This patch hides the problem again by expanding cursor lifetime until engine lives or another request is executed. As Artem noticed, this solution doesn't work in a common case and attempts to re-work this patch for a common case would be a way back to immortal cursors.

We need to understand the problem before fixing it. I believe we have a problem with sync of resulting data and arrow record batch lifetimes. Produced record batch can reference data it doesn't own (zero-copy conversion to arrow). When we get this batch, we don't need the cursor anymore and destroy it, so our data dies with the cursor leaving arrow record batch to reference nothing.

**Comment by igalink** [comment]
> Previously we didn't destroy cursors at all and the recent patch introduced their destruction.
Not this way. Before the fix, all cursors were destroyed along with the engine.

> This revealed some problems with data lifetime we don't understand.
Agree, but this issue requires studying the lifetime of C++ objects in python, which will take more time than this hotfix. And I don't see why this approach is bad: to manage the lifetime of an object from C++. If you want to destroy the object earlier, I can add the reset function to the interface.

> attempts to re-work this patch for a common case would be a way back to immortal cursors.
I don't see how this solution leads back to immortal cursors. Moreover tests show that cursors destroyed.

>I believe we have a problem with sync of resulting data and arrow record batch lifetimes. Produced record batch can reference >data it doesn't own (zero-copy conversion to arrow). When we get this batch, we don't need the cursor anymore and destroy >it, so our data dies with the cursor leaving arrow record batch to reference nothing.
Agree, I'll investigate this scenario. But still we'd better not give python the mission of deleting C ++ objects. Especially in the case of zero copy.

**Comment by ienkovich** [comment]
> Agree, but this issue requires studying the lifetime of C++ objects in python, which will take more time than this hotfix.
I believe you can easily reproduce the issue in C++ by destroying cursor and then working with remaining record batch. 

> And I don't see why this approach is bad: to manage the lifetime of an object from C++. If you want to destroy the object earlier, I can add the reset function to the interface.
You just prolongue object's lifetime for some random time you don't control (user controls it and doesn't even know this, he can currupt his data by simply executing another request). I believe that your original solution with unique_ptr is perfectly fine, the problem should be in another place.

> But still we'd better not give python the mission of deleting C ++ objects. Especially in the case of zero copy.
If we pass data from C++ to python then we have no way to understand its lifetime in C++, we should rely on Python object's lifetime. Again, your current solution with PyCursor should do the trick, C++ cursor lives while PyCursor object lives and dies after PyCursor object is destroyed. That's exactly what we need. We just need to make sure, that all required C++ data is properly referenced from Python objects we create.

**Comment by fexolm** [comment]
> > I mean, in the current solution, python would still own the Cursor and if you call reset, the cursor would be destructed on the python side. Isn't this a behaviour you would like to avoid for whatever reason?
> 
> Theoretically, this is so, but in practice we have a destroyed/corrupted cursor in tests.

I think we would better understand why the object corrupts first, and then create a fix. For now, I don't really understand what happands with python, and why we should store the object in c++.
But let's listen @ienkovich and @Garra1980 opinion.

**Pull request closed without merging**